out_datadog: add http proxy support

L3n41c opened the pull request:
Add a new plugin parameter, `proxy` that can be used to to specify an http proxy
to use to send datadog queries.

Signed-off-by: Lénaïc Huard <lenaic.huard@datadoghq.com>

L3n41c commented:
Travis build is broken because `apt` couldn’t connect to `ppa.launchpad.net`:
```
Err:42 http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu xenial/main amd64 libstdc++-7-dev amd64 7.4.0-1ubuntu1~16.04~ppa1
  Unable to connect to ppa.launchpad.net:http:
Err:43 http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu xenial/main amd64 g++-7 amd64 7.4.0-1ubuntu1~16.04~ppa1
  Unable to connect to ppa.launchpad.net:http:
Fetched 51.2 MB in 47s (1,082 kB/s)
E: Failed to fetch http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu/pool/main/g/gcc-9/liblsan0_9.2.1-17ubuntu1~16.04_amd64.deb  Could not connect to ppa.launchpad.net:80 (91.189.95.83), connection timed out
E: Failed to fetch http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu/pool/main/g/gcc-9/libtsan0_9.2.1-17ubuntu1~16.04_amd64.deb  Unable to connect to ppa.launchpad.net:http:
```

AppVeyor build is broken because `tests/internal/signv4.c` cannot find `unistd.h` on Windows:
```
  signv4.c
C:\projects\fluent-bit-2e87g\tests\internal\signv4.c(47): fatal error C1083: Cannot open include file: 'unistd.h': No such file or directory [C:\projects\fluent-bit-2e87g\build\tests\internal\flb-it-signv4.vcxproj]
```

That doesn’t seem to be linked to this change.

L3n41c commented:
Hello @edsiper !

The build of this PR is failing but I’m not sure it’s linked to my change.
The AppVeyor error seems to already be present in `master` (the miss of `unistd.h` on Windows)
And the Travis error seems to come from a transient network glitch that made apt unable to fetch packages from the repositories.
Would it be possible to relaunch the tests to check their results without the apt network error ?

Thanks!

clamoriniere commented:
Hi @edsiper 
can we have this PR merge for `1.4` ?

edsiper commented:
@clamoriniere 

a few changes will be required:

- please squash datadog changes in just one commit
- for other components changes do separate commits, e.g:
  - http_client: .... 
  - signv4: ......

clamoriniere commented:
@edsiper thank you the recommandation

@L3n41c could you make the changes?

edsiper commented on plugins/out_datadog/datadog_conf.c:
```diff
@@ -43,6 +43,46 @@ struct flb_out_datadog *flb_datadog_conf_create(struct flb_output_instance *ins,
     ctx->nb_additional_entries = 0;
 
+    tmp = flb_output_get_property("proxy", ins);
+    if (tmp) {
```
what about using flb_uri_create() here ?

https://github.com/fluent/fluent-bit/blob/master/src/flb_uri.c#L45

L3n41c replied:
This logic of parsing the `proxy` parameter is inherited from the `out_http` plugin:
https://github.com/fluent/fluent-bit/blob/master/plugins/out_http/http_conf.c#L61

Maybe this code snippet should be extracted from `http_conf.c` in order to be reused by `datadog_conf.c` but I’m not sure `flb_uri_create()` would do the job since the main goal of this code snippet is to extract the proxy host and tcp port whereas `flb_uri_create()` is for splitting an uri by `/`-separated tokens.

L3n41c replied:
`flb_utils_split_url()` will do the job!

edsiper commented on plugins/out_datadog/datadog_conf.c:
```diff
@@ -166,7 +206,19 @@ struct flb_out_datadog *flb_datadog_conf_create(struct flb_output_instance *ins,
     /* Prepare an upstream handler */
-    upstream = flb_upstream_create(config, ctx->host, ctx->port, io_flags, &ins->tls);
+    if (ctx->proxy) {
+        flb_trace("[out_datadog] Upstream Proxy=%s:%i",
```
please use new logging api: flb_plg_trace(ins, ...)

L3n41c replied:
Done.

edsiper commented on src/flb_http_client.c:
```diff
@@ -531,7 +531,7 @@ struct flb_http_client *flb_http_client(struct flb_upstream_conn *u_conn,
     char *fmt_plain =                           \
         "%s %s HTTP/1.%i\r\n";
     char *fmt_proxy =                           \
-        "%s http://%s:%i/%s HTTP/1.%i\r\n"
```
URIs always start with a ```/```, why adding this new one here ?

L3n41c replied:
In fact, I’m not adding a `/`, here. I’m removing it from `fmt_proxy` because `uri` is always prefixed by a `/`.
I think I got an issue because of the two consecutive slashes.

L3n41c commented:
@edsiper : I’ve extracted the fix for the proxy support in `flb_http_client` in a dedicated commit.
I’ve squashed the datadog related changes in another commit.
And I’ve added a 3rd commit to leverages `flb_utils_url_split()` to parse the `proxy` parameter to address the remark you raised on that part.

Let me know if you’d rather drop this 3rd commit or squash it.

L3n41c commented:
Hello @edsiper ,
Could you please have a new look at this PR and let me know if there are some other changes that you would like to be done ?

Thanks,
Lénaïc.

clamoriniere commented:
Hi @edsiper
I think @L3n41c has answered all your previous comments. Could you do a second review of this PR?
I hope it is not too late for the 1.4 RC, because we have users interested to the fluentbit with Datadog when this feature will be available.
Let me know if we can do something to ease your review.

Thank you in advance.
Cedric

edsiper commented:
we can merge it since I am still working on some improvements.

Please sanitize the commits, there is one commit that includes changes for two plugins:

- out_http, out_datadog: ...

split that in two commits.

L3n41c commented:
Hello @edsiper ,
I’ve split the commit that was touching the 2 plugins.

edsiper commented:
thanks!

L3n41c commented:
Just for the record, here is how I tested this PR:
* I started a squid proxy with this docker image: https://hub.docker.com/r/datadog/squid
* I then started fluentbit with the datadog proxy setting as described in fluent/fluent-bit-docs#249
* I checked that I received the logs on Datadog side and
* I checked in the squid logs that I could see the Datadog requests going through it.
